src/tracing/esp32_diagnostic_trace/Diagnostics.h

```diff
@@ -0,0 +1,249 @@
+     *
+     * @param writer A reference to the `chip::TLV::CircularTLVWriter` instance
+     *               used to encode the TLV data.
+     * @return CHIP_ERROR Returns an error code indicating the success or
+     *                    failure of the encoding operation.
+     */
+    virtual CHIP_ERROR Encode(chip::TLV::CircularTLVWriter & writer) const = 0;
+
+    virtual CHIP_ERROR Decode(chip::TLV::TLVReader & reader) = 0;
+};
+
+template <typename T>
+class Diagnostic : public DiagnosticEntry
+{
+public:
+    Diagnostic(char * label = nullptr, T value = T{}, uint32_t timestamp = 0) : label_(label), value_(value), timestamp_(timestamp)
+    {}
+
+    CHIP_ERROR Encode(chip::TLV::CircularTLVWriter & writer) const override
+    {
+        chip::TLV::TLVType DiagnosticOuterContainer = chip::TLV::kTLVType_NotSpecified;
+        ReturnErrorOnFailure(
+            writer.StartContainer(chip::TLV::AnonymousTag(), chip::TLV::kTLVType_Structure, DiagnosticOuterContainer));
+
+        // Write timestamp
+        ReturnErrorOnFailure(writer.Put(chip::TLV::ContextTag(DiagTag::TIMESTAMP), timestamp_));
+
+        // Write label
+        if (strlen(label_) > kMaxStringValueSize)
+        {
+            char labelBuffer[kMaxStringValueSize + 1];
+            memcpy(labelBuffer, label_, kMaxStringValueSize);
+            labelBuffer[kMaxStringValueSize] = '\0';
+            ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::LABEL), labelBuffer));
+        }
+        else
+        {
+            ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::LABEL), label_));
+        }
+
+        // Write value
+        if constexpr (std::is_same_v<T, char *>)
+        {
+            if (strlen(value_) > kMaxStringValueSize)
+            {
+                char valueBuffer[kMaxStringValueSize + 1];
+                memcpy(valueBuffer, value_, kMaxStringValueSize);
+                valueBuffer[kMaxStringValueSize] = '\0';
+                ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::VALUE), valueBuffer));
+            }
+            else
+            {
+                ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::VALUE), value_));
+            }
+        }
+        else if constexpr (std::is_same_v<T, uint32_t>)
+        {
+            ReturnErrorOnFailure(writer.Put(chip::TLV::ContextTag(DiagTag::VALUE), value_));
+        }
+        else if constexpr (std::is_same_v<T, int32_t>)
+        {
+            ReturnErrorOnFailure(writer.Put(chip::TLV::ContextTag(DiagTag::VALUE), value_));
+        }
+        else
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+        ReturnErrorOnFailure(writer.EndContainer(DiagnosticOuterContainer));
+        ReturnErrorOnFailure(writer.Finalize());
+        ChipLogProgress(DeviceLayer, "Diagnostic Value written to storage successfully. label: %s\n", label_);
```

Comment:
```suggestion
        ChipLogDetail(DeviceLayer, "Diagnostic Value written to storage successfully. label: %s\n", label_);
```

I think this may generate too much logs, so lets make this a debug print?